pkg/checks/traceroute/traceroute.go

```diff
@@ -2,197 +2,327 @@ package traceroute
 		wg.Add(1)
-		go func(t Target) {
-			l := log.With("target", t.Addr)
+		go func(ttl int) {
 			defer wg.Done()
-			l.Debug("Running traceroute")
-			start := time.Now()
-			trace, err := tr.traceroute(t.Addr, int(t.Port), int(tr.config.Timeout/time.Millisecond), tr.config.Retries, tr.config.MaxHops)
-			duration := time.Since(start)
+			err := helper.Retry(func(_ context.Context) error {
+				hop, err := traceroute(ctx, addr, ttl, timeoutDuration)
+				if hop != nil {
+					results <- *hop
+				}
+				if err != nil {
+					log.Error("traceroute failed", "err", err.Error(), "ttl", ttl)
+					return err
+				}
+				if !hop.Reached {
+					log.Debug("failed to reach target, retrying", "ttl", ttl)
+					return errors.New("failed to reach target")
+				}
+				return nil
+			}, cfg.Rc)(ctx)
 			if err != nil {
-				l.Error("Error running traceroute", "error", err)
+				log.Debug("traceroute could not reach target", "ttl", ttl)
 			}
+		}(ttl)
+	}
 
-			l.Debug("Ran traceroute", "result", trace, "duration", duration)
+	wg.Wait()
+	close(results)
 
-			r := result{
-				NumHops: len(trace.Hops),
-				Hops:    []hop{},
-			}
+	for r := range results {
+		hops[r.Ttl] = append(hops[r.Ttl], r)
+	}
 
-			for _, h := range trace.Hops {
-				r.Hops = append(r.Hops, hop{
-					Addr:    h.Host,
-					Latency: h.ElapsedTime,
-					Success: h.Success,
-				})
-			}
-			cResult <- internalResult{addr: t.Addr, res: r}
-		}(t)
+	printHops(ctx, hops)
+
+	return hops, nil
+}
+
+func ipFromAddr(remoteAddr net.Addr) net.IP {
+	switch addr := remoteAddr.(type) {
+	case *net.UDPAddr:
+		return addr.IP
+	case *net.TCPAddr:
+		return addr.IP
+	case *net.IPAddr:
+		return addr.IP
 	}
+	return nil
+}
 
-	log.Debug("Waiting for traceroute checks to finish")
+// traceroute performs a traceroute to the given address with the specified TTL and timeout.
+// It returns a Hop struct containing the latency, TTL, address, and other details of the hop.
+func traceroute(ctx context.Context, addr net.Addr, ttl int, timeout time.Duration) (*Hop, error) {
+	log := logger.FromContext(ctx)
+	canIcmp, icmpListener, err := setupIcmpListener()
+	if err != nil {
+		log.Error("Failed to open ICMP socket", "err", err.Error(), "ttl", ttl)
+		return nil, err
+	}
+	defer closeIcmpListener(canIcmp, icmpListener)
 
-	go func() {
-		wg.Wait()
-		close(cResult)
-	}()
+	start := time.Now()
+	conn, clientPort, err := tcpHop(addr, ttl, timeout)
+	latency := time.Since(start)
+	if err == nil {
+		return handleTcpSuccess(conn, addr, ttl, latency), nil
+	}
 
-	log.Debug("All traceroute checks finished")
+	if !canIcmp {
+		log.Debug("No permission for icmp socket", "ttl", ttl)
+		return &Hop{
+			Latency: latency,
+			Ttl:     ttl,
+			Reached: false,
+		}, nil
+	}
 
-	for r := range cResult {
-		res[r.addr] = r.res
+	h := handleIcmpResponse(ctx, icmpListener, clientPort, ttl, timeout)
+	h.Latency = latency
+	return &h, nil
+}
+
+func setupIcmpListener() (bool, *icmp.PacketConn, error) {
```

Comment:
Please rename this to `newICMPListener` to match the naming patterns of the sparrow.